File: scripts/verify.js
// 一次性验收（执行后退出，以退出码报告）：
//   阶段 1：针对不可整除约束直接调用求解器，确认无解与规范除尽障碍证据；
//   阶段 2：代码测试（node --test）；
//   阶段 3：构建检查（语法 + 模块加载）；
//   阶段 4：HTTP 冒烟——健康路径 + 本题复核接口（可解例 / 不可整除例 / 大整数例 / 取消）。
// 环境变量：BASE_URL 指向待验收服务（Compose 中为 http://web:3000）。
import { spawn } from 'node:child_process';
import { solveDiophantine } from '../src/diophantine.js';

const BASE_URL = process.env.BASE_URL || 'http://127.0.0.1:3000';
const failures = [];
const steps = [];

function step(name, fn) {
  steps.push({ name, fn });
}
function ok(msg) {
  console.log(`  ✓ ${msg}`);
}
function fail(msg) {
  failures.push(msg);
  console.error(`  ✗ ${msg}`);
}
function section(title) {
  console.log(`\n=== ${title} ===`);
}

function runCmd(cmd, args) {
  return new Promise((resolve) => {
    const p = spawn(cmd, args, { stdio: 'inherit' });
    p.on('exit', (code) => resolve(code));
  });
}

async function httpJson(path, options = {}, timeoutMs = 15000) {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const resp = await fetch(new URL(path, BASE_URL), { ...options, signal: ctrl.signal });
    const text = await resp.text();
    let body = null;
    try { body = JSON.parse(text); } catch { body = text; }
    return { status: resp.status, body };
  } finally {
    clearTimeout(timer);
  }
}

async function waitForReview(id, timeoutMs = 20000) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const { status, body } = await httpJson(`/api/reviews/${id}`);
    if (status !== 200) throw new Error(`轮询任务返回 HTTP ${status}`);
    if (body.status !== 'running') return body;
    await new Promise((r) => setTimeout(r, 150));
  }
  throw new Error('复核任务轮询超时');
}

// ---------- 阶段 1：不可整除约束的无解证据 ----------
step('不可整除约束：确认无解并给出规范除尽障碍', async () => {
  const A = [[2n, 1n, 1n], [1n, 2n, 2n], [3n, 3n, 3n]];
  const b = [4n, 10n, 14n];
  const r = await solveDiophantine(A, b);
  if (r.solvable) return fail('预期无解，但求解器报告有解');
  if (r.obstructions.length === 0) return fail('报告无解但未给出障碍证据');

  for (const o of r.obstructions) {
    // 证据可独立复核：c = u·b
    const c = o.uRow.reduce((s, u, i) => s + u * b[i], 0n);
    if (c !== o.transformedTarget) return fail(`证据 c≠u·b（${c} vs ${o.transformedTarget}）`);
    if (o.type === 'divisibility') {
      const d = o.pivot;
      if (d <= 0n) return fail('主元非正');
      if (o.uA.some((v) => v % d !== 0n)) return fail('u·A 存在不被主元整除的分量（证据不成立）');
      if (c % d === 0n) return fail('主元竟然整除变换后目标，矛盾结论错误');
      if (!(o.remainder > 0n && o.remainder < d)) return fail('规范余数越界');
      ok(`主元 d=${d} 不整除 c=${c}，规范余数 r=${o.remainder}（可逆行组合 u=[${o.uRow.join(',')}]）`);
    } else if (o.type === 'zero-row') {
      if (o.uA.some((v) => v !== 0n) || c === 0n) return fail('零行障碍证据不成立');
      ok(`零系数行对应非零目标 c=${c}`);
    }
  }
});

// ---------- 阶段 2：代码测试 ----------
step('代码测试（node --test）', async () => {
  const code = await runCmd(process.execPath, ['--test']);
  if (code !== 0) fail(`测试失败（退出码 ${code}）`);
});

// ---------- 阶段 3：构建检查 ----------
step('构建检查（语法 + 模块加载）', async () => {
  const code = await runCmd(process.execPath, ['scripts/build-check.js']);
  if (code !== 0) fail(`构建检查失败（退出码 ${code}）`);
});

// ---------- 阶段 4：HTTP 冒烟 ----------
step('HTTP 冒烟：健康路径 /healthz', async () => {
  const { status, body } = await httpJson('/healthz');
  if (status !== 200 || body.status !== 'ok') return fail(`健康路径异常：HTTP ${status}`);
  ok(`服务存活 ${BASE_URL}（uptime=${Math.round(body.uptime)}s）`);

  const home = await fetch(new URL('/', BASE_URL));
  if (home.status !== 200) return fail(`首页 HTTP ${home.status}`);
  const html = await home.text();
  if (!html.includes('垫片整数校正量复核')) return fail('首页内容不符');
  ok('首页可访问且内容正确');
});

step('HTTP 冒烟：可解复核返回精确整数校正量且每条约束平衡', async () => {
  const payload = {
    variables: ['shim_A', 'shim_B', 'shim_C'],
    matrix: [['2', '1', '-1'], ['1', '-3', '2'], ['3', '-2', '1']],
    targets: ['3', '5', '8'],
  };
  const created = await httpJson('/api/reviews', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (created.status !== 202 || !created.body.id) return fail(`发起复核失败：${created.status}`);
  const job = await waitForReview(created.body.id);
  if (job.status !== 'solvable') return fail(`预期 solvable，实际 ${job.status}`);
  const r = job.result;
  // 精确回代核验（用 BigInt 复核 HTTP 返回的文本）
  for (const row of r.constraints) {
    const s = row.terms.reduce((acc, t) => acc + BigInt(t.coefficient) * BigInt(t.value), 0n);
    if (s !== BigInt(row.target)) return fail(`约束 ${row.index + 1} 左侧和 ${s} ≠ 目标 ${row.target}`);
  }
  ok(`校正量 [${r.solution.join(', ')}]，全部 ${r.constraints.length} 条约束精确平衡`);
});

step('HTTP 冒烟：不可整除例返回无解与除尽障碍字段', async () => {
  const payload = {
    variables: ['shim_A', 'shim_B', 'shim_C'],
    matrix: [['2', '1', '1'], ['1', '2', '2'], ['3', '3', '3']],
    targets: ['4', '10', '14'],
  };
  const created = await httpJson('/api/reviews', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  });
  const job = await waitForReview(created.body.id);
  if (job.status !== 'unsolvable') return fail(`预期 unsolvable，实际 ${job.status}`);
  const o = job.result.obstructions[0];
  if (!o || (o.type !== 'divisibility' && o.type !== 'zero-row')) return fail('缺少障碍证据');
  if (!/^-?\d+$/.test(o.transformedTarget) || !/^-?\d+$/.test(o.pivot)) {
    return fail('障碍字段不是精确整数文本');
  }
  ok(`无解证据完整：类型=${o.type}，主元=${o.pivot}，变换后目标=${o.transformedTarget}，余数=${o.remainder}`);
});

step('HTTP 冒烟：超过安全整数范围的系数以精确文本呈现并正确求解', async () => {
  const huge = '1234567890123456789012345678901234567890'; // 40 位，远超 2^53
  // 由整数解 x0 = [2, -3] 反造目标，保证精确可解
  const x0 = [2n, -3n];
  const A = [[BigInt(huge), 1n], [0n, BigInt(huge)]];
  const b = A.map((row) => row[0] * x0[0] + row[1] * x0[1]);
  const payload = {
    variables: ['shim_X', 'shim_Y'],
    matrix: [[huge, '1'], ['0', huge]],
    targets: [b[0].toString(), b[1].toString()],
  };
  const created = await httpJson('/api/reviews', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  });
  const job = await waitForReview(created.body.id);
  if (job.status !== 'solvable') return fail(`大整数例预期 solvable，实际 ${job.status}（${JSON.stringify(job.error)}）`);
  if (job.result.solution[0] !== '2' || job.result.solution[1] !== '-3') {
    return fail(`大整数解不精确：${JSON.stringify(job.result.solution)}`);
  }
  // 返回的大系数必须是与输入逐位相同的文本
  const coeffText = job.result.constraints[0].terms[0].coefficient;
  if (coeffText !== huge) return fail(`大系数字符串失真：${coeffText}`);
  // 左侧和（含 40+ 位数字）必须逐位等于目标文本
  if (job.result.constraints[0].leftSum !== b[0].toString()) return fail('大整数左侧和失真');
  ok(`40 位系数逐位保留，校正量 [${job.result.solution.join(', ')}]，左侧和精确相等`);
});

step('HTTP 冒烟：非法输入（指数记法）被 400 拒绝', async () => {
  const payload = { variables: ['a'], matrix: [['1e3']], targets: ['0'] };
  const { status } = await httpJson('/api/reviews', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (status !== 400) return fail(`预期 400，实际 ${status}`);
  ok('拒绝了浮点/指数形式输入');
});

// ---------- 执行 ----------
(async () => {
  console.log(`验收目标: ${BASE_URL}`);
  for (let i = 0; i < steps.length; i++) {
    const { name, fn } = steps[i];
    section(`阶段 ${i + 1}/${steps.length}：${name}`);
    try {
      await fn();
    } catch (err) {
      fail(`${name} 抛出异常：${err.stack || err.message}`);
    }
  }

  section('验收结论');
  if (failures.length === 0) {
    console.log(`全部 ${steps.length} 个验收阶段通过。`);
    process.exit(0);
  }
  console.error(`${failures.length} 项验收失败：`);
  for (const f of failures) console.error(`  - ${f}`);
  process.exit(1);
})();

File: scripts/verify.py
#!/usr/bin/env python3
"""One-shot acceptance routine for the shim-correction review stack.

Order of checks (mirrors the acceptance contract):
  1. Confirm the no-solution evidence for a non-divisible constraint via
     the review API (with a target beyond the IEEE-754 safe-integer range).
  2. Run the code test-suite and the build checks.
  3. HTTP smoke-test the health path and the review endpoint.
The process exit code reports the overall result: 0 = pass, 1 = fail.
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
import time
import urllib.error
import urllib.request

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
APP_BASE_URL = os.environ.get("APP_BASE_URL", "http://127.0.0.1:8080").rstrip("/")

# 2**53 + 1 exceeds the JavaScript/IEEE-754 safe-integer range; the whole
# stack must still carry it as exact integer text.
BIG_TARGET = str(2**53 + 1)  # 9007199254740993

failures: list[str] = []


def check(name: str, condition: bool, detail: str = "") -> bool:
    print(f"[{'PASS' if condition else 'FAIL'}] {name}", flush=True)
    if not condition:
        if detail:
            print(f"       {detail}", flush=True)
        failures.append(name)
    return condition


def http(method: str, url: str, payload=None) -> tuple[int, str]:
    data = None
    headers = {}
    if payload is not None:
        data = json.dumps(payload).encode("utf-8")
        headers = {"Content-Type": "application/json"}
    request = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            return response.status, response.read().decode("utf-8")
    except urllib.error.HTTPError as exc:
        return exc.code, exc.read().decode("utf-8", errors="replace")
    except urllib.error.URLError as exc:
        return -1, str(exc)


def post_review(payload) -> tuple[int, dict | None]:
    status, text = http("POST", f"{APP_BASE_URL}/api/review", payload)
    try:
        return status, json.loads(text)
    except json.JSONDecodeError:
        return status, None


def wait_for_app(attempts: int = 30, delay: float = 1.0) -> bool:
    for _ in range(attempts):
        status, _ = http("GET", f"{APP_BASE_URL}/health")
        if status == 200:
            return True
        time.sleep(delay)
    return False


def step1_obstruction_evidence() -> None:
    print("\n== 步骤 1：不可整除约束的无解证据 ==", flush=True)
    status, body = post_review(
        {
            "variables": ["shim_north", "shim_south"],
            "matrix": [["2", "0"], ["0", "2"]],
            "target": [BIG_TARGET, "4"],
        }
    )
    check("不可整除约束复核返回 HTTP 200", status == 200, f"status={status}")
    check("判定为无整数解", bool(body) and body.get("solvable") is False,
          f"body={body!r}"[:400])
    obstruction = (body or {}).get("obstruction") or {}
    check("障碍类型为规范除尽障碍 non_divisible",
          obstruction.get("type") == "non_divisible", f"obstruction={obstruction!r}"[:400])
    check("主元精确为 2", obstruction.get("pivot") == "2")
    check("变换后目标以精确大整数文本呈现",
          obstruction.get("transformedTarget") == BIG_TARGET,
          f"got {obstruction.get('transformedTarget')!r}")
    check("余数精确为 1（主元不能整除变换后目标）",
          obstruction.get("remainder") == "1")
    u_terms = obstruction.get("uRowTerms") or []
    total = sum(int(term["product"]) for term in u_terms) if u_terms else None
    check("行变换各项乘积之和等于变换后目标",
          total is not None and str(total) == BIG_TARGET)


def run_subprocess(name: str, argv: list[str]) -> None:
    proc = subprocess.run(argv, cwd=ROOT, capture_output=True, text=True)
    output = (proc.stdout + proc.stderr).strip()
    check(name, proc.returncode == 0, output[-2000:] if proc.returncode != 0 else "")
    if proc.returncode == 0 and output:
        for line in output.splitlines()[-3:]:
            print(f"       {line}", flush=True)


def step2_tests_and_build_checks() -> None:
    print("\n== 步骤 2：代码测试与构建检查 ==", flush=True)
    run_subprocess(
        "单元测试（求解器与 HTTP 层）",
        [sys.executable, "-m", "unittest", "discover", "-s", "tests", "-t", "."],
    )
    run_subprocess(
        "语法构建检查 compileall",
        [sys.executable, "-m", "compileall", "-q", "app", "scripts", "tests"],
    )
    run_subprocess(
        "模块导入检查",
        [sys.executable, "-c", "import app.diophantine, app.server; print('imports ok')"],
    )
    for rel in (
        "Dockerfile",
        "docker-compose.yml",
        "app/static/index.html",
        "app/static/app.js",
        "app/static/style.css",
    ):
        check(f"构建所需文件存在：{rel}",
              os.path.exists(os.path.join(ROOT, rel)))


def step3_http_smoke() -> None:
    print("\n== 步骤 3：健康路径与复核接口 HTTP 冒烟 ==", flush=True)
    status, text = http("GET", f"{APP_BASE_URL}/health")
    healthy = False
    try:
        healthy = json.loads(text).get("status") == "ok"
    except json.JSONDecodeError:
        pass
    check("健康路径 /health 返回 200 且 status=ok", status == 200 and healthy,
          f"status={status} body={text[:200]}")

    status, text = http("GET", f"{APP_BASE_URL}/")
    check("页面 / 返回 200 且包含复核界面", status == 200 and "复核" in text,
          f"status={status}")

    status, body = post_review(
        {
            "variables": ["K1", "K2", "K3"],
            "matrix": [
                ["9007199254740993", "1", "0"],
                ["1", "3", "1"],
                ["0", "2", "4"],
            ],
            "target": ["18014398509481989", "12", "10"],
        }
    )
    check("复核接口（可解大整数用例）返回 HTTP 200", status == 200, f"status={status}")
    check("判定为可解", bool(body) and body.get("solvable") is True,
          f"body={body!r}"[:400])
    check("精确整数校正量为 [2, 3, 1]",
          bool(body) and body.get("solution") == ["2", "3", "1"])
    constraints = (body or {}).get("constraints") or []
    check("每条约束的左侧和均精确等于目标值",
          bool(constraints) and all(c["satisfied"] for c in constraints))
    if constraints:
        first_products = [t["product"] for t in constraints[0]["terms"]]
        check("首条约束各项乘积精确（含超大整数）",
              first_products == [str(9007199254740993 * 2), "3", "0"],
              f"products={first_products!r}")

    status, body = post_review(
        {"variables": ["x"], "matrix": [[1.5]], "target": [1]}
    )
    check("浮点系数被拒绝（HTTP 400，绝不四舍五入）",
          status == 400 and bool(body) and body.get("ok") is False,
          f"status={status} body={body!r}"[:300])


def main() -> int:
    print(f"验收目标：{APP_BASE_URL}", flush=True)
    if not wait_for_app():
        check("等待应用就绪", False, f"{APP_BASE_URL}/health 未在限定时间内就绪")
    else:
        print("应用已就绪。", flush=True)
        step1_obstruction_evidence()
        step2_tests_and_build_checks()
        step3_http_smoke()

    print("\n== 验收结论 ==", flush=True)
    if failures:
        print(f"失败 {len(failures)} 项：", flush=True)
        for name in failures:
            print(f"  - {name}", flush=True)
        return 1
    print("全部验收项通过。", flush=True)
    return 0


if __name__ == "__main__":
    sys.exit(main())
